### PR TITLE
datatrails: fix bookmark/recent trail detection, prevent duplications, save trail on browser close/reload

### PR DESCRIPTION
**Pull request by darrenjaneczek**
**What is this feature?**

- fixes detection on bookmarks and recents when current step isn't the last step on the trail history
- now saves current trail on browser close or reload (unload)
- refresh page or return to recent trail URL will restore state to that trail and allow user to resume
  that trail instead of creating a duplicate recent trail
- do not create a recent trail out of an empty starting trail

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by torkelo — APPROVED** [review]
Nice tests! 

The bookmark button state should cleared after any change is made as the bookmark is only for that specific state. And if possible clear the history when we create the bookmark.

**Comment by darrenjaneczek** [comment]
@torkelo 
> Nice tests!
> 
> The bookmark button state should cleared after any change is made as the bookmark is only for that specific state. And if possible clear the history when we create the bookmark.

Resolved the bookmark button state being kept after changing state via:
- https://github.com/grafana/grafana/pull/85677/commits/01ba40b2da31d30c728e1aac6676a47e7da642d1

Merging in the fix from:
- https://github.com/grafana/grafana/pull/86843 
also helps with bookmark stability by not retroactively changing the parameter state of steps.

I had a local stash saved to only store the bookmark state vs. the whole history. It is not in the following draft PR is here for consideration:
- https://github.com/grafana/grafana/pull/86929

File: public/app/features/trails/DataTrailsApp.tsx [hunk 1]
```diff
@@ -99,8 +91,23 @@ let dataTrailsApp: DataTrailsApp;
 
 export function getDataTrailsApp() {
   if (!dataTrailsApp) {
+    const newTrail = newMetricsTrail();
+
+    // Set the initial state of the newTrail based on the URL,
+    // In case we are initializing from an externally created URL or a page reload
+    getUrlSyncManager().initSync(newTrail);
+    // Remove the URL sync for now. It will be restored on the trail if it is activated.
+    getUrlSyncManager().cleanUp(newTrail);
```
Comment (torkelo) [inline public/app/features/trails/DataTrailsApp.tsx:100]: is this really needed?

Comment (darrenjaneczek) [reply public/app/features/trails/DataTrailsApp.tsx:100]: Yes, in order to find the "recentMatchingTrail" below.

This is part of resolving the "continue same trail on refresh" bug.

**Review by torkelo — APPROVED** [review]
Looks ok.

Still confused why we store the full state in bookmark vs just the URL but a separate topic :)

File: public/app/features/trails/DataTrailsApp.tsx (hunk 1 above)
Comment (darrenjaneczek) [reply public/app/features/trails/DataTrailsApp.tsx:100]: (unless there is a more straightforward way of populating the trail state from the current url)

Comment (torkelo) [reply public/app/features/trails/DataTrailsApp.tsx:100]: @darrenjaneczek continue same trail on refresh? I refresh should clear the history

Comment (darrenjaneczek) [reply public/app/features/trails/DataTrailsApp.tsx]: It was deemed to be a bug.
With this change, accessing the same URL, either by refresh, or explicit URL, will act similar to selecting the trail from the recent list.

**Pull request merged**